### PR TITLE
Allow Fakified subclass to have different device for inner and outer tensor

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141842
* #141840
* #141841
* __->__ #141839
* #141166

Previously if a wrapper tensor subclass is fakified, the inner tensors would end up having the same device as the outer tensor. This PR makes it so that inner and outer tensors can have different devices. 

See OffloadTensor PR https://github.com/pytorch/pytorch/pull/141840/files#diff-3bc0cf540b694f4ec0a3749f78b047456657a53a5657e495ffb68e5970c5fdaaR1955 for an application. A simpler test has been added in this PR.

This is technically bc-breaking because now the callback passed to MetaConverter needs to accept an extra argument, but no one external should be using this anyway?

### CODE REVIEW

File: torch/_subclasses/meta_utils.py [hunk 1]
```diff
@@ -756,10 +757,11 @@ def meta_tensor(
         self,
         t: MetaTensorDesc,
         shape_env: Optional[ShapeEnv],
-        callback: Callable[[Callable[[], torch.Tensor]], _TensorT],
+        callback_: Callable[[Callable[[], torch.Tensor]], _TensorT],
```
Comment (bdhirsh) [inline torch/_subclasses/meta_utils.py:760]: maybe a dumb nit but any reason not to keep the arg name here the same? (and do `callback = functools.partial(callback, device=t.device)`). I agree this isn't public API though

**Review by bdhirsh — APPROVED** [review]
lgtm, can you add a test?

File: torch/_subclasses/meta_utils.py (hunk 1 above)
Comment (soulitzer) [reply torch/_subclasses/meta_utils.py]: Yep you're right `callback = functools.partial(callback, device=t.device)` should just overwrite it

**Pull request closed without merging**